src/blockchain/Blockchain.ts

```diff
@@ -141,28 +211,130 @@ export default class Blockchain implements IBlockchainApi {
+   *
+   * Transaction fees will apply whenever a transaction fee makes it into a block, even if extrinsics fail to execute correctly!
+   *
+   * @param identity The [[Identity]] to potentially re-sign the Tx with.
+   * @param tx The [[SubmittableExtrinsic]] to be submitted. Most transactions need to be signed, this must be done beforehand.
+   * @param opts Partial optional criteria for resolving/rejecting the promise.
+   * @returns A promise which can be used to track transaction status.
+   * If resolved, this promise returns the eventually resolved [[SubmittableResult]].
+   */
+  public async submitSignedTx(
+    identity: Identity,
+    tx: SubmittableExtrinsic,
+    opts?: Partial<SubscriptionPromiseOptions>
+  ): Promise<SubmittableResult> {
+    const options = parseSubscriptionOptions(opts)
+    const retry = async (reason: Error): Promise<SubmittableResult> => {
+      if (reason.message === ERROR_TRANSACTION_RECOVERABLE().message) {
+        return submitSignedTx(await this.reSignTx(identity, tx), options)
+      }
+      throw reason
+    }
+    return submitSignedTx(tx, options).catch(retry).catch(retry)
+  }
+
+  /**
+   * [ASYNC] Signs and submits the SubmittableExtrinsic with optional resolution and rejection criteria.
+   *
+   * @param identity The [[Identity]] that we sign and potentially re-sign the tx with.
+   * @param tx The generated unsigned [[SubmittableExtrinsic]] to submit.
+   * @param opts Partial optional criteria for resolving/rejecting the promise.
+   * @returns Promise result of The Extrinsic.
+   *
+   */
   public async submitTx(
     identity: Identity,
     tx: SubmittableExtrinsic,
-    opts: SubscriptionPromiseOptions = {}
+    opts?: Partial<SubscriptionPromiseOptions>
   ): Promise<SubmittableResult> {
     const signedTx = await this.signTx(identity, tx)
-    return Blockchain.submitSignedTx(signedTx, opts)
+    return this.submitSignedTx(identity, signedTx, opts)
+  }
+
+  /**
+   * [ASYNC] Retrieves the Nonce for Transaction signing for the specified account and increments the in accountNonces mapped Index.
+   *
+   * @param accountAddress The address of the identity that we retrieve the nonce for.
+   * @returns [[BN]] representation of the Tx nonce for the identity.
+   *
+   */
+  public async getNonce(accountAddress: string): Promise<BN> {
+    let nonce = this.accountNonces.get(accountAddress)
+    if (!nonce) {
+      const chainNonce = await this.api.rpc.system
+        .accountNextIndex(accountAddress)
+        .catch((reason) => {
+          log.error(
+            `On-chain nonce retrieval failed for account ${accountAddress}\nwith reason: ${reason}`
+          )
+          throw Error(`Chain failed to retrieve nonce for : ${accountAddress}`)
+        })
+      const secondQuery = this.accountNonces.get(accountAddress)
+      nonce = BN.max(chainNonce, secondQuery || new BN(0))
+    }
+    this.accountNonces.set(accountAddress, nonce.addn(1))
+    return nonce
```

Comment:
Now that i revisited that code, I notice that it's not clear why we do the things we do. I would maybe add some comments

```suggestion
  public async getNonce(accountAddress: string): Promise<BN> {
    let nonce = this.accountNonces.get(accountAddress)
    if (!nonce) {
      // the account nonce is unknown, we will query it from chain
      const chainNonce = await this.api.rpc.system
        .accountNextIndex(accountAddress)
        .catch((reason) => {
          log.error(
            `On-chain nonce retrieval failed for account ${accountAddress}\nwith reason: ${reason}`
          )
          throw Error(`Chain failed to retrieve nonce for : ${accountAddress}`)
        })
      // ensure that the nonce we queried is still up to date and no newer nonce was queried during the await above
      const secondQuery = this.accountNonces.get(accountAddress)
      nonce = BN.max(chainNonce, secondQuery || new BN(0))
    }
    this.accountNonces.set(accountAddress, nonce.addn(1))
    return nonce
```